File: 9_first_Duplicate_Array/first.js
// First Duplicate Value - Google - Interview Question -
// in JavaScript

function firstDuplicate(array) {
    const arrLength = array.length;
    if (arrLength === 1) return -1;
    if (arrLength === 0) return -1;

    let valuesMap = new Map();

    for (let index = 0; index < arrLength; index += 1) {
        let valueAtIndex = array[index]
        let valueIndexes = valuesMap.get(valueAtIndex);

        if (valueIndexes == undefined || !valueIndexes) {
            valuesMap.set(valueAtIndex, [index]);
        } else {
            valueIndexes.push(index);
            if (valueIndexes.length == 2)
                return valueAtIndex;
        };
    }
    return -1;
}

console.log(firstDuplicate([2, 1, 3, 5, 3, 2])); // 3

File: 9_first_Duplicate_Array/first.go
package main

import "fmt"

// First Duplicate Value - Google - Interview Question -
// in Golang
func firstDuplicate(array []int) int {
	arrLength := len(array)
	if arrLength == 0 || arrLength == 1 {
		return -1
	}

	var valuesMap = map[int][]int{}

	for index := 0; index < arrLength; index++ {
		valueAtIndex := array[index]
		valueIndexes, mapRefExist := valuesMap[valueAtIndex]

		if !mapRefExist {
			valuesMap[valueAtIndex] = []int{index}
		} else {
			valueIndexes = append(valueIndexes, index)
			if len(valueIndexes) == 2 {
				return valueAtIndex
			}
		}
	}
	return -1
}

func main() {
	fmt.Println(firstDuplicate([]int{2, 1, 3, 5, 3, 2})) //3
}
